File: task/src/main/java/name/shokred/popugjira/task/entity/TaskEntity.java
package name.shokred.popugjira.task.entity;

import lombok.*;

import javax.persistence.*;
import java.time.ZonedDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder

@Table(name = "TASKS")
@Entity
public class TaskEntity {

    @Id
    @Column(name = "TASK_ID")
    private Long id;

    @Column(name = "DESCRIPTION")
    private String description;

    @Column(name = "STATUS")
    @Enumerated(value = EnumType.STRING)
    private TaskStatus status;

    public void setDescription(String description) {
        setChangeDate(ZonedDateTime.now());
        this.description = description;
    }

    public void setStatus(TaskStatus status) {
        setChangeDate(ZonedDateTime.now());
        this.status = status;
    }

    @Column(name = "CHANGE_DATE")
    private ZonedDateTime changeDate;

    public boolean canChangeTo(TaskStatus statusTo) {
        return statusTo.isAvailableFrom(this.status);
    }
}

File: task/persistence/src/main/kotlin/name/shokred/popug_jira/task/entity/TaskEntity.kt
package name.shokred.popug_jira.task.entity

import name.shokred.popug_jira.task.Task
import javax.persistence.Entity
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.Table

@Entity
@Table(name = "TASKS")
class TaskEntity(
    var title: String,
    var description: String,
    var status: String,
    @ManyToOne
    @JoinColumn(name = "user_id")
    var assignee: Long?
) : BaseEntity<Long>() {

    fun toDomain(): Task {
        val task = Task(id!!, title, description)
        if (status == "COMPLETED") {
            task.complete()
        }
        return task
    }

    companion object {
        fun fromDomain(task: Task): TaskEntity {
            return TaskEntity(
                task.title,
                task.description,
                if (task.isCompleted()) "COMPLETED" else "NOT_COMPLETED",
                task.assignee?.id
            )
                .apply { id = task.id }
        }
    }
}
